src/neuroglancer/datasource/computed/backend.ts

```diff
@@ -0,0 +1,594 @@
+    const gridKey = gridPositionKey(originChunk.chunkGridPosition);
+    this.originGridPositions_.delete(gridKey);
+
+    const chunkSize = originChunk.chunkDataSize!;
+    const numChannels = inputSpec.numChannels;
+    const chunkCorner = vec3.multiply(
+        vec3.create(), originChunk.chunkGridPosition, this.source.originSource.spec.chunkDataSize);
+
+    let destination = getArrayView(this.inputBuffer_!, inputSpec.dataType);
+
+    const source = this.maybeDecodeBuffer_(
+        <TypedArray>(originChunk.data!), inputSpec.dataType, chunkSize, numChannels);
+
+    copyBufferOverlap(
+        chunkCorner, chunkSize, source, this.inputLower_!, inputSpec.size, destination,
+        inputSpec.dataType);
+    originChunk.unregisterListener(this);
+    setTimeout(() => this.checkDone_(), 0);
+  }
+
+  /**
+   * Peforms the computation over the input buffer, ensuring validity of the
+   * eventual output data that will be set for this chunk. This includes
+   * handling volume-boundary effects.
+   */
+  private performComputation_() {
+    if (this.cancellationToken_!.isCanceled) {
+      return Promise.reject(CANCELED);
+    }
+    const computation = this.source.computation;
+    const {outputSpec} = this.computationParams_!;
+    const outputSize = outputSpec.size;
+    const outputDataType = outputSpec.dataType;
+
+    // Most of the time, the chunk data size corresponds to the output buffer
+    // size, but a chunk at the upper bound of a volume will be clipped to the
+    // volume bounds. Computations are guaranteed the same buffer sizes each
+    // time, so we check for this situation and perform a crop-and-copy when
+    // necessary.
+    return computation.compute(this.inputBuffer_!, this.cancellationToken_!)
+        .then((outputBuffer) => {
+          this.inputBuffer_ = undefined;
+          if (vec3.equals(outputSize, this.chunkDataSize!)) {
+            decodeRawChunk(this, outputBuffer);
+            return;
+          }
+          const outputBufferView = getArrayView(outputBuffer, outputDataType);
+          const chunkBuffer = new ArrayBuffer(
+              prod(this.chunkDataSize!) * outputSpec.numChannels * DATA_TYPE_BYTES[outputDataType]);
+          const chunkBufferView = getArrayView(chunkBuffer, outputDataType);
+          const outputCorner = vec3.multiply(vec3.create(), this.chunkGridPosition, outputSize);
+          copyBufferOverlap(
+              outputCorner, outputSize, outputBufferView, outputCorner, this.chunkDataSize!,
+              chunkBufferView, outputDataType);
+          decodeRawChunk(this, chunkBuffer);
+        });
+  }
+
+  /**
+   * Idempotently performs the computation, if the input buffer is ready. This
+   * function should be called after a timeout in most cases, because it may
+   * take a long time to return.
+   */
+  private checkDone_() {
+    if (this.computing_) {
+      return;
+    }
+    if (this.originGridPositions_.size === 0) {
+      this.computing_ = true;
```

Comment:
I think this.cleanup() could actually be called here, since the cleanup it does is only related to copying of origin chunks, not the computation.

Reply:
Done.